### DISCUSSION TITLE
Frequency and VoltageSource

### DISCUSSION THREAD

**Discussion by termostatos** [SpiceSharp/SpiceSharp#208] [Q&A] [answered]
Hi, first of all thank you for all this job.
I'm been using SpiceShapr some time to simulate DC circuits and now I'm starting with AC circuits. The only problem that I have is that I don't know how to asign the frequency to a VoltageSource.  
Thanks in advance.

`

           var ckt = new Circuit(
               new VoltageSource("V1", "in", "0", 0).SetParameter("acmag",27500),
               new Resistor("R1", "in", "1", 0.1),
               new Inductor("l1", "1","2", 0.1),
               new Resistor("R2", "2", "3", 2),
               new Inductor("l2", "3", "4", 0.5),
               new Resistor("R3", "4", "5", 0.01),
               new Inductor("l3", "5", "0", 0.1)
               );

            // Create the simulation
            var tran = new Transient("Tran 1", 1, 10,10);

            // Make the exports
            var inputExport = new RealVoltageExport(tran, "in");

            // Simulate
            tran.ExportSimulationData += (sender, args) =>
            {
                var input = inputExport.Value;
                //var output = outputExport.Value;

                Console.WriteLine("V in V1: " + input  );
            };
            tran.Run(ckt);
`

**Comment by marcin-golebiowski** [comment]
Please take a look at:
https://spicesharp.github.io/SpiceSharp/articles/tutorial/acanalysis.html

For Transient you can also use **Sine** waveform for VoltageSource.
Take a look at:
 - https://github.com/SpiceSharp/SpiceSharp/blob/master/SpiceSharpTest/Waveforms/PwlTests.cs
 - https://github.com/SpiceSharp/SpiceSharp/blob/master/SpiceSharpTest/Waveforms/SineTests.cs
 - https://github.com/SpiceSharp/SpiceSharp/blob/master/SpiceSharpTest/Waveforms/PulseTests.cs
 - https://github.com/SpiceSharp/SpiceSharp/tree/dace7bddaadf7ac22bc32abd2dc61b4d28afb13d/SpiceSharp/Components/Waveforms/Sine

**Comment by svenboulanger** [comment, marked as answer]
Hi!

Marcin already mentioned in his post, but here's another summary.

You can only choose a phase and a magnitude for a voltage source, and the frequency will be decided by the AC simulation. An AC analysis does a sweep in frequency.

I however see that you are not defining an AC analysis, but a `Transient` analysis. If you want to see a sine wave for that type of simulation, then you need to use the `Sine()` waveform on the voltage source.

```csharp
// ...
new VoltageSource("V1", "in", "0", new Sine(0, 1, 27500)),
// ...
```

Yours sincerely,
Sven

**Reply by termostatos** [reply]
Thank you very much!!
That's was I want and it works.

But now i have a problem and it's that a want the current of a resistor and if I do the RealCurrentExport on a Resistor the value is 0. Code below. I try to make de ComplexCurrentExport but is only for AC analys.

```
            var ckt = new Circuit(
                           new VoltageSource("V1", "in", "0", new Sine(0, 50, 27500)),
                            new Resistor("R1", "in", "1", 0.01),
                           new Inductor("l1", "1", "2", 0.1),
                           new Resistor("MainR", "2", "3", 2),
                           new Resistor("R3", "3", "4", 0.01),
                           new Inductor("l3", "4", "0", 0.1)
                           );
            
            // Create the simulation
            var tran = new Transient("Tran", 1,10,10);

            // Make the exports
            var inputExport = new RealVoltageExport(tran, "2");
            var inputCurrentExport = new RealCurrentExport(tran, "MainR");
            

            // Simulate
            tran.ExportSimulationData += (sender, args) =>
            {
                var input = inputExport.Value;
                var current = inputCurrentExport.Value;

                Console.WriteLine("V in V1: " + input + "\n" +
                    "Current in the Main Resistor is: " + current);
            };
            tran.Run(ckt);
        }
```

Thanks in advance and Happy new year!!

**Reply by svenboulanger** [reply]
One thing I couldn't help but notice is that your transient parameters seem to be orders of magnitude different from the voltage source sine wave frequency. Perhaps you should try a time step of around 1us? A coarse timestep may fail to capture behaviors at a much shorter timescale.

**Reply by termostatos** [reply]
Oh, yes, that could be a problem. I'm trying to reply this circuit in LT Spice:
![imagen](https://user-images.githubusercontent.com/33329368/211292182-47b9161a-ba73-49ca-90ef-afb9ecb31fce.png)

So, do you know what are the appropriate parameters for the _new Transient()_ function?

**Reply by svenboulanger** [reply]
Did you try the one I wrote in my message?
